File: scripts/differential/linuxforhealth_oracle.py
#!/usr/bin/env python
"""Read side of the differential harness for its second oracle, LinuxForHealth x12.

Two subcommands, each writing one JSON object to stdout:

  describe <icvn>  the installed distribution's name, version and declared
                   licence, the interpreter it runs under, every distribution
                   its runtime requirements pull in with the version actually
                   installed, and every transaction model the distribution
                   carries under its `v5010` package whose implementation guide
                   belongs to that interchange control version.
  read             one X12 document, read from stdin and put through the
                   transaction model for the implementation guide the document
                   declares, then returned as the segment stream that model
                   consumed.

Nothing here interprets an X12 rule and nothing here names a model: the models
are enumerated from the installed package, and the model a document goes
through is the one the oracle's own `X12ModelReader` selects from the
document's own GS-08 (falling back to ST-03). A document that model refuses is
reported as a refusal carrying the exception kind, and never as a reading,
because comparing a document the guide model rejected on a guide-agnostic split
would be comparing against a reader that did not read it as the guide.
"""

from __future__ import annotations

import importlib
import importlib.metadata
import inspect
import json
import platform
import pkgutil
import re
import sys
from typing import Any, Iterator

PACKAGE = "linuxforhealth-x12"
MODEL_PACKAGE = "linuxforhealth.x12.v5010"
MODEL_MODULE = re.compile(r"^x12_([0-9A-Z]{3})_([0-9]{6}X[0-9A-Z]+)$")
REQUIREMENT_NAME = re.compile(r"^\s*([A-Za-z0-9][A-Za-z0-9._-]*)")


def _normalise(name: str) -> str:
    """The distribution name as PEP 503 compares it."""
    return re.sub(r"[-_.]+", "-", name).lower()


def _licence() -> dict[str, str | None]:
    """The licence the installed distribution declares about itself."""
    metadata = importlib.metadata.metadata(PACKAGE)
    classifiers = [c for c in (metadata.get_all("Classifier") or []) if c.startswith("License ::")]
    return {
        "declared": metadata.get("License"),
        "expression": metadata.get("License-Expression"),
        "classifier": classifiers[0] if classifiers else None,
    }


def _companions() -> list[dict[str, str | None]]:
    """Every distribution the oracle's runtime requirements pull in, as installed.

    The walk starts at the oracle's own `Requires-Dist` and follows each
    requirement that is not behind an `extra` marker. A requirement behind any
    other marker is followed when its distribution is installed, so the list
    can only grow past what the resolver needed, never shrink below it. A
    requirement that names a distribution that is not installed is listed with
    no version, which the harness refuses.
    """
    seen: dict[str, str | None] = {}
    pending = [PACKAGE]
    while pending:
        current = pending.pop()
        try:
            requirements = importlib.metadata.requires(current) or []
        except importlib.metadata.PackageNotFoundError:
            continue
        for requirement in requirements:
            head, _, marker = requirement.partition(";")
            if "extra" in marker:
                continue
            match = REQUIREMENT_NAME.match(head)
            if match is None:
                continue
            name = _normalise(match.group(1))
            if name in seen:
                continue
            try:
                version: str | None = importlib.metadata.version(name)
            except importlib.metadata.PackageNotFoundError:
                if marker.strip():
                    continue
                version = None
            seen[name] = version
            if version is not None:
                pending.append(name)
    return [{"package": name, "version": seen[name]} for name in sorted(seen)]


def _bindings(icvn: str) -> list[dict[str, str | None]]:
    """The transaction models the installed distribution carries for one ICVN."""
    from linuxforhealth.x12 import parsing

    package = importlib.import_module(MODEL_PACKAGE)
    bindings = []
    for module in sorted(pkgutil.iter_modules(package.__path__), key=lambda m: m.name):
        match = MODEL_MODULE.match(module.name)
        if not module.ispkg or match is None:
            continue
        transaction_set, guide = match.group(1), match.group(2)
        if not guide.startswith(icvn):
            continue
        # The oracle's own resolution from (transaction set, guide) to the
        # transaction model class, so the model named here is the one `read`
        # puts a document through.
        model = parsing._load_transaction_model(transaction_set, guide)
        doc = inspect.getdoc(model) if model is not None else None
        bindings.append(
            {
                "icvn": icvn,
                "vriic": guide,
                "transactionSet": transaction_set,
                "fic": None,
                "tspc": None,
                "model": f"{MODEL_PACKAGE}.{module.name}",
                "modelId": model.__name__ if model is not None else None,
                "modelTitle": doc.splitlines()[0].strip() if doc else None,
            }
        )
    return bindings


def describe(icvn: str) -> dict[str, Any]:
    return {
        "package": PACKAGE,
        "version": importlib.metadata.version(PACKAGE),
        "licence": _licence(),
        "python": platform.python_version(),
        "companions": _companions(),
        "bindingSource": MODEL_PACKAGE,
        "interchangeControlVersion": icvn,
        "bindings": _bindings(icvn),
    }


def read(text: str) -> dict[str, Any]:
    from linuxforhealth.x12.io import X12ModelReader

    consumed: list[list[str]] = []
    try:
        with X12ModelReader(text) as reader:
            segment_reader = reader._x12_segment_reader
            framed = segment_reader.segments

            def recording() -> Iterator[tuple[str, list[str]]]:
                # Record every segment exactly as the oracle framed it, on its
                # way into the transaction model, so the stream returned is the
                # one the model read and not a second, independent split.
                for name, fields in framed():
                    consumed.append(list(fields))
                    yield name, fields

            segment_reader.segments = recording
            delimiters = segment_reader.delimiters
            for _model in reader.models():
                pass
    except Exception as err:  # noqa: BLE001 - every refusal is reported with its kind
        return {"ok": False, "refusal": {"kind": type(err).__name__, "detail": str(err)}}
    return {
        "ok": True,
        "delimiters": {
            "element": delimiters.element_separator,
            "component": delimiters.component_separator,
            "repetition": delimiters.repetition_separator,
            "segment": delimiters.segment_terminator,
        },
        "segments": [{"id": fields[0], "elements": fields[1:]} for fields in consumed],
    }


def main(argv: list[str]) -> int:
    if len(argv) >= 2 and argv[0] == "describe":
        json.dump(describe(argv[1]), sys.stdout)
        return 0
    if len(argv) >= 1 and argv[0] == "read":
        json.dump(read(sys.stdin.read()), sys.stdout)
        return 0
    sys.stderr.write(
        "usage: linuxforhealth_oracle.py describe <icvn> | linuxforhealth_oracle.py read < document\n"
    )
    return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: scripts/differential/linuxforhealth-oracle.ts
/**
 * The second oracle: LinuxForHealth x12, an independent open-source X12 reader
 * with a transaction model per implementation guide, driven out of process
 * through `scripts/differential/linuxforhealth_oracle.py`.
 *
 * SECURITY: every subprocess is array-form `spawnSync`, no shell.
 *
 * Like the first oracle it is a development-only tool and never a dependency of
 * the published package. Unlike the first, its own metadata leaves a package it
 * cannot run without free: it declares Pydantic with a floor and no ceiling
 * while importing a Pydantic 1 API, so an unpinned resolve picks a Pydantic it
 * cannot import. Its requirement therefore pins, exactly, the oracle, every
 * companion package its runtime requirements pull in, and the interpreter
 * version it runs under. Two routes obtain it from that one requirement:
 *
 *   - by default, `uv run --python <version> --with <pin>...`, which resolves
 *     and caches the exact set for this invocation and leaves nothing
 *     installed;
 *   - when `X12_DIFFERENTIAL_LINUXFORHEALTH_PYTHON` names an interpreter, that
 *     interpreter, for an environment that already provisioned the oracle.
 *
 * Neither route is trusted to have obtained the right thing. The oracle reports
 * its own version, the interpreter version it runs under and every companion
 * package actually installed, and this module refuses an answer that differs
 * from the pin anywhere, and a companion the installation shows that the
 * requirement does not pin.
 */

import { dirname, join } from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

import {
  OracleUnavailableError,
  type CorpusDocument,
  type DifferentialOracle,
  type OracleBinding,
  type OracleDescription,
  type OracleRead,
} from "./harness.js";
import {
  invokeOracle,
  normaliseDistribution,
  OracleRequirementError,
  parseExactPin,
  type ExactPin,
} from "./oracle-process.js";

const ORACLE_SCRIPT = join(dirname(fileURLToPath(import.meta.url)), "linuxforhealth_oracle.py");

/** The distribution name this driver reads. */
export const LINUXFORHEALTH_PACKAGE = "linuxforhealth-x12";

/** Environment variable naming an interpreter that already has the oracle. */
export const LINUXFORHEALTH_INTERPRETER_ENV = "X12_DIFFERENTIAL_LINUXFORHEALTH_PYTHON";

/** The requirement as the command line states it, before anything is checked. */
export interface LinuxForHealthRequirement {
  /** `linuxforhealth-x12==<version>`. */
  readonly oracle: string;
  /** One `<name>==<version>` per companion package. */
  readonly companions: readonly string[];
  /** The interpreter version, `<major>.<minor>.<micro>`. */
  readonly python: string | null;
}

/** The requirement once every part of it is known to be exact. */
export interface LinuxForHealthPin {
  readonly label: string;
  readonly oracle: ExactPin;
  readonly companions: readonly ExactPin[];
  readonly python: string;
}

/**
 * Check the requirement pins everything exactly, before anything is invoked:
 * the oracle, each companion, and the interpreter version.
 */
export function parseLinuxForHealthRequirement(
  requirement: LinuxForHealthRequirement,
): LinuxForHealthPin {
  const label = requirement.oracle;
  const oracle = parseExactPin(requirement.oracle, label);
  if (normaliseDistribution(oracle.name) !== LINUXFORHEALTH_PACKAGE) {
    throw new OracleRequirementError(
      label,
      `This driver reads ${LINUXFORHEALTH_PACKAGE}, and the requirement names "${oracle.name}".`,
    );
  }
  const companions = requirement.companions.map((companion) => parseExactPin(companion, label));
  const seen = new Set<string>([LINUXFORHEALTH_PACKAGE]);
  for (const companion of companions) {
    const name = normaliseDistribution(companion.name);
    if (seen.has(name)) {
      throw new OracleRequirementError(
        label,
        `The requirement for ${label} pins "${companion.name}" more than once, or pins the oracle ` +
          `itself as a companion.`,
      );
    }
    seen.add(name);
  }
  if (requirement.python === null || !/^\d+\.\d+\.\d+$/u.test(requirement.python)) {
    throw new OracleRequirementError(
      label,
      `The requirement for ${label} must pin an exact interpreter version as ` +
        `"<major>.<minor>.<micro>", got ${JSON.stringify(requirement.python)}.`,
    );
  }
  return { label, oracle, companions, python: requirement.python };
}

/** The argv that obtains and invokes the oracle, for one subcommand. */
export function linuxforhealthInvocation(
  pin: LinuxForHealthPin,
  args: readonly string[],
): string[] {
  const interpreter = process.env[LINUXFORHEALTH_INTERPRETER_ENV];
  if (interpreter !== undefined && interpreter !== "") {
    return [interpreter, ORACLE_SCRIPT, ...args];
  }
  return [
    "uv",
    "run",
    "--isolated",
    "--python",
    pin.python,
    "--with",
    `${pin.oracle.name}==${pin.oracle.version}`,
    ...pin.companions.flatMap((companion) => ["--with", `${companion.name}==${companion.version}`]),
    "--",
    "python",
    ORACLE_SCRIPT,
    ...args,
  ];
}

interface RawDescription {
  package: string;
  version: string;
  licence: { declared: string | null; expression: string | null; classifier: string | null };
  python: string;
  companions: { package: string; version: string | null }[];
  bindingSource: string;
  interchangeControlVersion: string;
  bindings: OracleBinding[];
}

/**
 * Hold what the installation reports against the pin, refusing any difference.
 * Returns the companions as installed, in a stable order.
 */
function checkInstallation(
  pin: LinuxForHealthPin,
  raw: RawDescription,
  invocation: string,
): { package: string; version: string }[] {
  const unavailable = (detail: string): OracleUnavailableError =>
    new OracleUnavailableError(pin.label, invocation, detail);
  if (raw.version !== pin.oracle.version) {
    throw unavailable(
      `the oracle reports version ${raw.version}, and this harness is pinned to ` +
        `${pin.oracle.version}. A report produced against a different revision of the oracle ` +
        `would describe a reader nobody can reproduce.`,
    );
  }
  if (raw.python !== pin.python) {
    throw unavailable(
      `the oracle runs under Python ${raw.python}, and this harness is pinned to ${pin.python}.`,
    );
  }
  const pinned = new Map(
    pin.companions.map((companion) => [normaliseDistribution(companion.name), companion.version]),
  );
  const installed: { package: string; version: string }[] = [];
  for (const companion of raw.companions) {
    const name = normaliseDistribution(companion.package);
    if (companion.version === null) {
      throw unavailable(`the oracle requires ${name}, which is not installed.`);
    }
    const version = pinned.get(name);
    if (version === undefined) {
      throw new OracleRequirementError(
        pin.label,
        `The installed ${LINUXFORHEALTH_PACKAGE} runs on ${name} ${companion.version}, and the ` +
          `requirement for ${pin.label} pins no version for it, so the resolver is free to move it.`,
      );
    }
    if (version !== companion.version) {
      throw unavailable(
        `the oracle runs on ${name} ${companion.version}, and this harness is pinned to ${version}.`,
      );
    }
    installed.push({ package: name, version: companion.version });
  }
  const runsOn = new Set(installed.map((companion) => companion.package));
  for (const name of pinned.keys()) {
    if (!runsOn.has(name)) {
      throw new OracleRequirementError(
        pin.label,
        `The requirement for ${pin.label} pins ${name}, which the installed ` +
          `${LINUXFORHEALTH_PACKAGE} does not run on, so the report would record a pin that fixed nothing.`,
      );
    }
  }
  return installed.sort((a, b) => a.package.localeCompare(b.package));
}

/**
 * The LinuxForHealth-backed oracle. Throws {@link OracleRequirementError} at
 * once when the requirement leaves anything free.
 */
export function linuxforhealthOracle(
  requirement: LinuxForHealthRequirement,
  icvn: string,
): DifferentialOracle {
  const pin = parseLinuxForHealthRequirement(requirement);
  // Each call settles its promise, a refusal included, rather than throwing
  // past it.
  return {
    describe(): Promise<OracleDescription> {
      return new Promise((resolve) => {
        const argv = linuxforhealthInvocation(pin, ["describe", icvn]);
        const raw = invokeOracle(pin.label, argv) as RawDescription;
        const companions = checkInstallation(pin, raw, argv.join(" "));
        resolve({
          package: raw.package,
          version: raw.version,
          licence: raw.licence.expression ?? raw.licence.declared ?? "unstated",
          licenceClassifier: raw.licence.classifier,
          bindingKind: "model",
          bindingSource: raw.bindingSource,
          interchangeControlVersion: raw.interchangeControlVersion,
          python: raw.python,
          companions,
          bindings: raw.bindings,
        });
      });
    },
    read(document: CorpusDocument): Promise<OracleRead> {
      return new Promise((resolve) => {
        const argv = linuxforhealthInvocation(pin, ["read"]);
        resolve(invokeOracle(pin.label, argv, document.text) as OracleRead);
      });
    },
  };
}
